### DISCUSSION TITLE
Is there a way to disable the `pages` functionality in Nextjs 15 RC

### DISCUSSION THREAD

**Discussion by robahtou** [vercel/next.js#66649] [Help]
### Summary

We're beginning to upgrade to Nextjs 15 RC to see what we need to update (we are on v14@latest). One of the things we'd like to use in our `src` folder is a folder called `pages`. We have our routes listed using the app router (folder/files) and that works great. And from a code organizing POV we'd like to create a folder called `pages` and add components there as well. (without of course having those components (folder/files) in any way automatically used for routing).

Currently, if we add the `pages` folder and add  an `index.tsx` or a `MyComponent/index.tsx`, they are also treaded as a route and will be rendered as a page in the browser.

We also attempted to use `pageExtensions: ['page.tsx', 'page.ts', 'page.jsx', 'page.js']` in the config file but then we'll have to update ALL our files in the app router directory as well.

Is there a way to just disable the `pages` functionality and have it be a 'dumb' folder?

**Comment by jdhiro** [comment]
Just ran into this as well, migrating an app from Vite to Next. The existing router uses the pages directory, and I spent about a day debugging why it wouldn't build!

Since we aren't using Next pages router, it would nice to disable it.

**Comment by nikhcevov** [comment]
I've just renamed `pages` to `pagez`, so it works for now. 

Btw we need a proper solution, as there is app folder for routing purposes, it would be great to disable pages router and special rules for `pages` folder

**Comment by DevanandGowda** [comment]
Ended up renaming the `pages` folder to `screens`, I was following this convention on one of my Expo projects. It still irks me a little bit though.

**Comment by x-tahosin** [comment]
Two clean options, depending on where you want the folder:

**Option 1 — Put it inside `app/` with an underscore prefix (recommended):**

App Router treats any folder whose name starts with `_` as a **private folder** — it's opted out of routing entirely, while still being importable from your app code:

```
src/
  app/
    _pages/           ← not a route, just a folder
      Header.tsx
      Sidebar.tsx
    page.tsx
    layout.tsx
```

You can then import `from "@/app/_pages/Header"` without Next.js trying to make `/_pages` a route.

Docs: https://nextjs.org/docs/app/getting-started/project-structure#private-folders

**Option 2 — Put it outside `app/` altogether:**

The `pages/` directory is *only* treated as the Pages Router when it's at the project root (or `src/pages` if you use `src`). Put your component folder anywhere else and it's just a folder:

```
src/
  app/
  components/pages/   ← also fine, no underscore needed
    Header.tsx
```

The `pageExtensions` workaround can technically work (if you set it to something Pages-Router files don't match, like `['page.tsx', 'page.ts']`, only files named exactly `page.tsx`/`page.ts` become routes — your `pages/MyComponent/index.tsx` is then ignored). But yeah, it's noisy and easy to break if anyone later adds a `page.tsx` inside the wrong folder.

For your use case I'd go with **Option 1** (`app/_pages/`) — the only catch is that it has to be `_pages` (with the underscore) rather than literally `pages`. The underscore is what tells the App Router to skip routing for that folder. If renaming is a no-go and you really need the literal name `pages`, **Option 2** (move it outside `app/`) is the way.
